fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricCodecDataProvider.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.datagen.v1.provider;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+import com.google.gson.JsonElement;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.JsonOps;
+
+import net.minecraft.data.DataOutput;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.DataWriter;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator;
+import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
+
+/**
+ * Extend this class and implement {@link FabricCodecDataProvider#configure}.
+ *
+ * <p>Register an instance of the class with {@link FabricDataGenerator.Pack#addProvider} in a {@link net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint}.
+ */
+public abstract class FabricCodecDataProvider<T> implements DataProvider {
+	private final DataOutput.PathResolver pathResolver;
+	private final Codec<T> codec;
+
+	protected FabricCodecDataProvider(FabricDataOutput dataOutput, DataOutput.OutputType outputType, String directoryName, Codec<T> codec) {
+		this.pathResolver = dataOutput.getResolver(outputType, directoryName);
+		this.codec = codec;
+	}
+
+	@Override
+	public CompletableFuture<?> run(DataWriter writer) {
+		Map<Identifier, JsonElement> entries = new HashMap<>();
+		BiConsumer<Identifier, T> provider = (id, value) -> {
+			JsonElement json = this.convert(id, value);
+			JsonElement existingJson = entries.put(id, json);
+
+			if (existingJson != null) {
+				throw new IllegalArgumentException("Duplicate entry " + id);
+			}
+		};
+
+		this.configure(provider);
+		return this.write(writer, entries);
+	}
+
+	protected abstract void configure(BiConsumer<Identifier, T> provider);
+
+	private JsonElement convert(Identifier id, T value) {
+		DataResult<JsonElement> dataResult = this.codec.encodeStart(JsonOps.INSTANCE, value);
+		return dataResult.result().orElseThrow(() -> new IllegalStateException("Invalid entry " + id));
```

Comment:
Might be good to not swallow the original error message by using `DataResult.get` instead:
```suggestion
		return dataResult.get()
						.mapRight(partial -> "Invalid entry %s: %s".formatted(id, partial.message()))
						.orThrow();
```
(Written in github comments so not 100% sure whether it compiles; also `RuntimeException` is better here since `IllegalStateException` has a different purpose that doesn't apply here)

Reply:
Finally had the time to look at the review. I thought I saw this exception being thrown in other places too, hence why I used it for consistency, but I must've misread it since I can't seem to find it anymore.